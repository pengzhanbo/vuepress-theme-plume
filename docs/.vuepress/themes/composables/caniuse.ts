import type { ComputedRef, Ref } from 'vue'
import { onClickOutside, useDebounceFn, useEventListener, useLocalStorage } from '@vueuse/core'
import { computed, onMounted, readonly, ref, watch } from 'vue'

interface Feature {
  label: string
  value: string
}

interface SelectItem {
  label: string
  value: string
}

const api = 'https://caniuse.pengzhanbo.cn/features.json'

const pastVersions: SelectItem[] = [
  { label: '不显示旧版本', value: '0' },
  ...Array.from({ length: 5 }).fill(0).map((_, i) => ({
    label: `旧版本（当前版本 - ${i + 1}）`,
    value: `${i + 1}`,
  })),
]

const futureVersions: SelectItem[] = [
  { label: '不显示未来版本', value: '0' },
  ...Array.from({ length: 3 }).fill(0).map((_, i) => ({
    label: `未来版本（当前版本 + ${i + 1}）`,
    value: `${i + 1}`,
  })),
]

const embedTypes: SelectItem[] = [
  { label: 'iframe', value: '' },
  { label: 'image', value: 'image' },
]

export function useCaniuseVersionSelect(): {
  past: Ref<string>
  future: Ref<string>
  embedType: Ref<string>
  pastList: Readonly<SelectItem[]>
  futureList: Readonly<SelectItem[]>
  embedTypeList: Readonly<SelectItem[]>
} {
  const past = ref('2')
  const future = ref('1')
  const embedType = ref('')

  const pastList = readonly(pastVersions)
  const futureList = readonly(futureVersions)
  const embedTypeList = readonly(embedTypes)

  return {
    past,
    future,
    pastList,
    futureList,
    embedType,
    embedTypeList,
  }
}

export function useCaniuseFeaturesSearch(
  inputEl: Ref<HTMLInputElement | null>,
  listEl: Ref<HTMLUListElement | null>,
): {
    featureList: Ref<Feature[] | undefined>
    isFocus: Ref<boolean>
    feature: ComputedRef<string>
    onSelect: (item: Feature) => void
  } {
  const features = useLocalStorage('plume:caniuse-feature-list', [] as Feature[])
  const featuresUpdated = useLocalStorage('plume:caniuse-feature-list-updated', Date.now())
  const maxAge = 1000 * 60 * 60 * 24 * 3 // 3 days
  onMounted(async () => {
    if (typeof document === 'undefined')
      return

    if (features.value.length && Date.now() - featuresUpdated.value < maxAge)
      return
    const data = await fetch(api).then(res => res.json())
    features.value = data || features.value || []
  })

  const input = ref('')
  const isFocus = ref(false)
  const searched = ref<Feature[]>()

  const selected = ref<Feature | null>(null)

  watch(() => [features.value, isFocus.value], () => {
    if (!isFocus.value)
      searched.value = features.value
  }, { immediate: true })

  onClickOutside(listEl, () => {
    isFocus.value = false
  }, { ignore: [inputEl] })

  useEventListener(inputEl, 'input', useDebounceFn(() => {
    selected.value = null
    input.value = inputEl.value?.value || ''

    if (!input.value) {
      searched.value = features.value
    }
    else {
      searched.value = features.value.filter(item => item.label.includes(input.value) || item.value.includes(input.value))
      if (searched.value.length === 1)
        selected.value = searched.value[0]
    }
  }, 500))

  useEventListener(inputEl, 'focus', () => {
    isFocus.value = true
  })

  function onSelect(item: Feature): void {
    selected.value = item
    isFocus.value = false
    if (inputEl.value)
      inputEl.value.value = item.label
  }

  return {
    featureList: searched,
    isFocus,
    onSelect,
    feature: computed(() => selected.value?.value || ''),
  }
}

export function useCaniuse({ feature, embedType, past, future }: {
  feature: Ref<string>
  embedType: Ref<string>
  past: Ref<string>
  future: Ref<string>
}): {
    output: ComputedRef<string>
    rendered: ComputedRef<string>
  } {
  const output = computed(() => {
    let content = '@[caniuse'
    if (embedType.value)
      content += ` ${embedType.value}`

    if (past.value !== '-2' || future.value !== '1') {
      if (past.value === '0' && future.value === '0')
        content += '{0}'
      else
        content += `{-${past.value},${future.value}}`
    }

    content += ']('

    if (feature.value)
      content += feature.value

    return `${content})`
  })

  const rendered = computed(() => {
    if (!feature.value || !embedType.value)
      return ''
    return resolveCanIUse(feature.value)
  })

  return { output, rendered }
}

function resolveCanIUse(feature: string): string {
  const link = 'https://caniuse.bitsofco.de/image/'
  const alt = `Data on support for the ${feature} feature across the major browsers from caniuse.com`
  return `<p><picture>
    <source type="image/webp" srcset="${link}${feature}.webp">
    <source type="image/png" srcset="${link}${feature}.png">
    <img src="${link}${feature}.jpg" alt="${alt}" width="100%">
  </picture></p>`
}
