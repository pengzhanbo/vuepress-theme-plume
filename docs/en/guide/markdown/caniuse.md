---
title: Can I Use
createTime: 2025/10/08 14:50:55
icon: streamline:desktop-help
permalink: /en/guide/markdown/caniuse/
---

## Overview

When writing articles, this feature provides the functionality to embed platform support information for WEB features from [can-i-use](https://caniuse.com/).

This allows for a more intuitive representation of a feature's support level when describing a specific WEB feature.

## Configuration

This feature is disabled by default. You can enable it in the configuration file.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdownP: {
      caniuse: true, // [!code ++]
    },
  })
})
```

In your article's markdown file, use the following format:

``` md
@[caniuse](feature)
```

For ease of use, the theme provides tool support: [caniuse Feature Search](../../../tools/caniuse.md).
You can use this tool directly to help generate the markdown code.

## Syntax

``` md
@[caniuse](feature)
@[caniuse{browser_versions}](feature)
@[caniuse embed_type](feature)
@[caniuse embed_type{browser_versions}](feature)
```

- `feature`

   Required. For correct values, please refer to [caniuse-embed.vercel.app](https://caniuse-embed.vercel.app/zh-CN).

- `{browser_versions}`

  Optional. Specifies the support status of the current feature across multiple browser versions.

  Default value: `{-2,1}`

  Format: `{past,future}` Value range: `-5 ~ 3`

  - Values less than `0` indicate support status for browser versions lower than the current one.
  - `0` indicates the support status for the current browser version.
  - Values greater than `0` indicate support status for browser versions higher than the current one.

- `embed_type`

  Optional. The type of resource embedding.

  Type: `'embed' | 'image'`

  Default value: `'embed'`

:::caution
The `image` type is no longer recommended. It is advised to use the `embed` type.
The theme has changed the implementation technology for the embed component.
The current `embed` type now offers significant advantages over the `image` type,
including faster loading speed, smaller size, and better interactive experience.
:::

## Examples

**Get the browser support for the CSS pseudo-class selector `:dir()`:**

```md
@[caniuse](css-matches-pseudo)
```

Result:

@[caniuse](css-matches-pseudo)

**Get the browser support for the CSS pseudo-class selector `:dir()` as an image:**

```md
@[caniuse image](css-matches-pseudo)
```

Result:

@[caniuse image](css-matches-pseudo)

**Get the browser support for the CSS pseudo-class selector `:dir()` for a specific range of browser versions:**

```md
@[caniuse{-2,3}](css-matches-pseudo)
```

Result:

@[caniuse{-2,3}](css-matches-pseudo)
