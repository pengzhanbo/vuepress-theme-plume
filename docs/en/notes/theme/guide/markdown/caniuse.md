---
title: Can I Use
createTime: 2025/03/24 22:10:32
icon: streamline:desktop-help
permalink: /en/guide/markdown/caniuse/
---

## Overview

When writing articles, you can embed the support status of [can-i-use](https://caniuse.com/) WEB features across platforms.

This makes it easier to describe the support level of a particular WEB feature.

## Configuration

This feature is not enabled by default. You can enable it in the configuration file.

```ts title=".vuepress/config.ts"
export default defineUserConfig({
  theme: plumeTheme({
    markdownP: {
      caniuse: true, // [!code ++]
    },
  })
})
```

In your Markdown file, use the following format:

```md
@[caniuse](feature)
```

For convenience, the theme provides a tool: [caniuse feature search](../../../../../notes/tools/caniuse.md), which can help generate the Markdown code.

## Syntax

```md
@[caniuse](feature)
@[caniuse{browser_versions}](feature)
@[caniuse embed_type](feature)
@[caniuse embed_type{browser_versions}](feature)
```

- `feature`

  Required. For correct values, refer to [caniuse-embed.vercel.app](https://caniuse-embed.vercel.app/zh-CN)

- `{browser_versions}`

  Optional. The support status of the feature across multiple browser versions.

  Default value: `{-2,1}`

  Format: `{past,future}` with values ranging from `-5 ~ 3`

  - Less than `0` indicates support below the current browser version
  - `0` indicates support at the current browser version
  - Greater than `0` indicates support above the current browser version

- `embed_type`

  Optional. The type of resource embedding.

  Type: `'embed' | 'image'`

  Default value: `'embed'`

:::caution
The use of image type is no longer recommended. Instead, use the embed type, as the theme has changed the embed implementation, offering faster loading, smaller size, and better interaction.
:::

## Examples

**Get the browser support status for the CSS pseudo-class selector `:dir()`:**

```md
@[caniuse](css-matches-pseudo)
```

Effect:

@[caniuse](css-matches-pseudo)

**Get the browser support status for the CSS pseudo-class selector `:dir()` as an image:**

```md
@[caniuse image](css-matches-pseudo)
```

Effect:

@[caniuse image](css-matches-pseudo)

**Get the browser support status for the CSS pseudo-class selector `:dir()` for specific browser versions:**

```md
@[caniuse{-2,3}](css-matches-pseudo)
```

Effect:

@[caniuse{-2,3}](css-matches-pseudo)
