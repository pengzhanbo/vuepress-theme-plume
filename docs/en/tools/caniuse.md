---
title: CanIUse Feature Search
icon: fa-brands:css3
createTime: 2025/10/09 12:41:26
permalink: /en/tools/caniuse/
readingTime: false
editLink: false
---

In addition to using this tool, you can also directly visit [caniuse.com/](https://caniuse.com/) to search for features.
On the results page, click the `#` button on the left, and the browser's address bar will update automatically. For example:

Search for `@property`, click `#`, and it will navigate to `https://caniuse.com/mdn-css_at-rules_property`.
You can directly copy `mdn-css_at-rules_property` and paste it into your markdown file:

```md
@[caniuse](mdn-css_at-rules_property)
```

## Tool

This tool helps search for features in caniuse and generates the corresponding caniuse markdown code.

<script setup>
import CanIUseConfig from '@source/.vuepress/themes/components/CanIUseConfig.vue'
</script>

<CanIUseConfig />
