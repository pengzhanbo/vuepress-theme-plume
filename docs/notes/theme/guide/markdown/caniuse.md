---
title: Can I Use
createTime: 2024/09/30 14:50:55
icon: streamline:desktop-help
permalink: /guide/markdown/caniuse/
---

## 概述

在编写文章时， 提供嵌入 [can-i-use](https://caniuse.com/) WEB feature 各平台支持说明 的功能。

方便在描述某个 WEB feature 时，能更直观的表述 该特性的支持程度。

## 配置

此功能默认不启用，你可以在配置文件中启用它。

::: code-tabs
@tab .vuepress/config.ts

```ts
export default defineUserConfig({
  theme: plumeTheme({
    plugins: {
      markdownPower: {
        caniuse: true, // [!code highlight]
      },
    }
  })
})
```

:::

在你的 文章 markdown文件中，使用以下格式：

``` md
@[caniuse](feature)
```

为了方便使用，主题提供了工具支持：[caniuse 特性搜索](../../../tools/caniuse.md)，你可以直接使用该工具
帮助生成 markdown 代码。

## 语法

``` md
@[caniuse](feature)
@[caniuse{browser_versions}](feature)
@[caniuse embed_type](feature)
@[caniuse embed_type{browser_versions}](feature)
```

- `feature`

   必填。 正确取值请参考 [caniuse-embed.vercel.app](https://caniuse-embed.vercel.app/zh-CN)

- `{browser_versions}`

  可选。当前特性在多个版本中的支持情况。

  默认值为: `{-2,1}`

  格式： `{past,future}`  取值范围为 `-5 ~ 3`

  - 小于`0` 表示低于当前浏览器版本的支持情况
  - `0` 表示当前浏览器版本的支持情况
  - 大于`0` 表示高于当前浏览器版本的支持情况

- `embed_type`

  可选。 资源嵌入的类型。

  类型： `'embed' | 'image'`

  默认值为：`'embed'`

:::caution
不再推荐使用 image 类型，建议使用 embed 类型，主题更换了 embed 实现技术方案，
现在的 embed 类型优势已远远超过 image 类型，加载速度更快，体积更小，交互体验更好。
:::

## 示例

**获取 css 伪类选择器 `:dir()` 在各个浏览器的支持情况：**

```md
@[caniuse](css-matches-pseudo)
```

效果：

@[caniuse](css-matches-pseudo)

**以图片的形式，获取 css 伪类选择器 `:dir()` 在各个浏览器的支持情况：**

```md
@[caniuse image](css-matches-pseudo)
```

效果：

@[caniuse image](css-matches-pseudo)

**获取 css 伪类选择器 `:dir()` 特定范围浏览器的支持情况：**

```md
@[caniuse{-2,3}](css-matches-pseudo)
```

效果：

@[caniuse{-2,3}](css-matches-pseudo)
