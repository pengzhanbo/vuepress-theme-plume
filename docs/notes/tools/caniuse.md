---
title: caniuse 特性搜索
author: pengzhanbo
icon: fa-brands:css3
createTime: 2024/04/16 12:41:26
permalink: /tools/caniuse/
readingTime: false
editLink: false
---

::: info 提示
除了使用本工具，你也可以直接访问 [caniuse.com/](https://caniuse.com/) 搜索特性，
在结果页中，点击 左侧的 `#` 按钮后，浏览器的地址栏会自动跳转，例如：

搜索 `@property`,点击 `#` 会跳转到 `https://caniuse.com/mdn-css_at-rules_property`，
可以直接复制 `mdn-css_at-rules_property` ，粘贴到 markdown 文件中:

```md :no-line-numbers
@[caniuse](mdn-css_at-rules_property)
```

:::

本工具用于帮助搜索 caniuse 中的特性。并生成 caniuse markdown 代码。

<CanIUseConfig />
