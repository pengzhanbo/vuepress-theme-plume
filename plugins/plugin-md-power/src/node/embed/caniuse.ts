/**
 * @[caniuse embed{1,2,3,4}](feature_name)
 * @[caniuse image](feature_name)
 */
import type { PluginWithOptions } from 'markdown-it'
import type MarkdownIt from 'markdown-it'
import type Token from 'markdown-it/lib/token.mjs'
import type { CanIUseMode, CanIUseOptions, CanIUseTokenMeta } from '../../shared/index.js'
import container from 'markdown-it-container'
import { customAlphabet } from 'nanoid'
import { createEmbedRuleBlock } from './createEmbedRuleBlock.js'

const nanoid = customAlphabet('abcdefghijklmnopqrstuvwxyz', 5)
const UNDERLINE_RE = /_+/g
/**
 * @example
 * ```md
 * @[caniuse](feature_name)
 * ```
 */
export const caniusePlugin: PluginWithOptions<CanIUseOptions> = (
  md,
  { mode: defaultMode = 'embed' }: CanIUseOptions = {},
): void => {
  createEmbedRuleBlock<CanIUseTokenMeta>(md, {
    type: 'caniuse',
    syntaxPattern: /^@\[caniuse\s*(embed|image)?(?:\{([0-9,\-]*)\})?\]\(([^)]*)\)/,
    meta: ([, mode, versions = '', feature]) => ({
      feature,
      mode: (mode as CanIUseMode) || defaultMode,
      versions,
    }),
    content: meta => resolveCanIUse(meta),
  })
}

/**
 * @deprecated use caniuse plugin
 *
 * 兼容旧语法
 * @example
 * ```md
 * :::caniuse <feature_name>
 * :::
 * ```
 */
export function legacyCaniuse(
  md: MarkdownIt,
  { mode = 'embed' }: CanIUseOptions = {},
): void {
  const modeMap: CanIUseMode[] = ['image', 'embed']
  const isMode = (mode: CanIUseMode): boolean => modeMap.includes(mode)

  mode = isMode(mode) ? mode : modeMap[0]
  const type = 'caniuse'
  const validateReg = new RegExp(`^${type}`)

  const validate = (info: string): boolean => {
    return validateReg.test(info.trim())
  }

  const render = (tokens: Token[], index: number): string => {
    const token = tokens[index]
    if (token.nesting === 1) {
      const info = token.info.trim().slice(type.length).trim() || ''
      const feature = info.split(/\s+/)[0]
      const versions = info.match(/\{(.*)\}/)?.[1] || ''
      return feature ? resolveCanIUse({ feature, mode, versions }) : ''
    }
    else {
      return ''
    }
  }

  md.use(container, type, { validate, render })
}

function resolveCanIUse({ feature, mode, versions }: CanIUseTokenMeta): string {
  if (!feature)
    return ''

  if (mode === 'image') {
    const link = 'https://caniuse.bitsofco.de/image/'
    const alt = `Data on support for the ${feature} feature across the major browsers from caniuse.com`
    return `<ClientOnly><p><picture>
      <source type="image/webp" srcset="${link}${feature}.webp">
      <source type="image/png" srcset="${link}${feature}.png">
      <img src="${link}${feature}.jpg" alt="${alt}" width="100%">
    </picture></p></ClientOnly>`
  }

  feature = feature.replace(UNDERLINE_RE, '_')
  const { past, future } = resolveVersions(versions)
  const meta = nanoid()

  return `<CanIUseViewer feature="${feature}" meta="${meta}" past="${past}" future="${future}" />`
}

function resolveVersions(versions: string): { past: number, future: number } {
  if (!versions)
    return { past: 2, future: 1 }

  const list = versions
    .split(',')
    .map(v => Number(v.trim()))
    .filter(v => !Number.isNaN(v) && v >= -5 && v <= 3)

  list.push(0)

  const uniq = [...new Set(list)].sort((a, b) => b - a)
  return {
    future: uniq[0],
    past: Math.abs(uniq[uniq.length - 1]),
  }
}
