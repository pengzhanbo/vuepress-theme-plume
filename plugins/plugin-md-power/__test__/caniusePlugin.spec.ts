import type { CanIUseOptions } from '../src/shared/index.js'
import MarkdownIt from 'markdown-it'
import { beforeEach, describe, expect, it, vi } from 'vitest'
import { caniusePlugin, legacyCaniuse } from '../src/node/embed/caniuse.js'

beforeEach(() => {
  vi.mock('../src/node/utils/nanoid.js', () => ({
    nanoid: vi.fn(() => 'test-id'),
  }))
})

function createMarkdown(options?: CanIUseOptions) {
  return MarkdownIt().use((md) => {
    md.block.ruler.before('code', 'import_code', () => false)
    md.renderer.rules.import_code = () => ''
  }).use(caniusePlugin, options)
}

describe('caniusePlugin', () => {
  it('should work', () => {
    const md = createMarkdown()

    expect(md.render('@[caniuse](feature)')).toMatchSnapshot()
    expect(md.render('@[caniuse image](feature)')).toMatchSnapshot()
    expect(md.render('@[caniuse embed](feature)')).toMatchSnapshot()
    expect(md.render('@[caniuse {-2,4}](feature)')).toMatchSnapshot()
    expect(md.render(`\
@[caniuse {-2,4}](feature)
@[caniuse {-2,}](feature)
`)).toMatchSnapshot()
    expect(md.render('@[caniuse embed{-2,4}](feature)')).toMatchSnapshot()
  })

  it('should work with options', () => {
    const md = createMarkdown({ mode: 'image' })

    expect(md.render('@[caniuse](feature)')).toMatchSnapshot()
    expect(md.render('@[caniuse embed](feature)')).toMatchSnapshot()
    expect(md.render('@[caniuse embed{-2,4}](feature)')).toMatchSnapshot()
  })

  it('should not work', () => {
    const md = createMarkdown()

    expect(md.render('@[caniuse]()')).toMatchSnapshot()
    expect(md.render('@[caniuse]xxx')).toMatchSnapshot()
    expect(md.render('@[ caniuse]()')).toMatchSnapshot()
    expect(md.render('@[caniuse](xxx')).toMatchSnapshot()
    expect(md.render('@[caniuse](')).toMatchSnapshot()
    expect(md.render('@[caniuse')).toMatchSnapshot()
    expect(md.render('@[caniuse)]()')).toMatchSnapshot()
    expect(md.render('@[caniuseee)]()')).toMatchSnapshot()
  })
})

describe('legacyCaniuse', () => {
  it('should work', () => {
    const md = MarkdownIt()
    legacyCaniuse(md)

    expect(md.render(':::caniuse feature\n:::')).toMatchSnapshot()
    expect(md.render(':::caniuse feature{-2,4}\n:::')).toMatchSnapshot()
    expect(md.render(':::caniuse\n:::')).toMatchSnapshot()
  })

  it('should work with unknown mode', () => {
    const md = MarkdownIt()
    legacyCaniuse(md, { mode: 'unknown' as any })

    expect(md.render(':::caniuse feature\n:::')).toMatchSnapshot()
    expect(md.render(':::caniuse feature{-2,4}\n:::')).toMatchSnapshot()
  })
})
