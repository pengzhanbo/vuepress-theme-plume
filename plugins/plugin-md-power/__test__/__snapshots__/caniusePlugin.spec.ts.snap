// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`caniusePlugin > should not work 1`] = `""`;

exports[`caniusePlugin > should not work 2`] = `
"<p>@[caniuse]xxx</p>
"
`;

exports[`caniusePlugin > should not work 3`] = `
"<p>@<a href=""> caniuse</a></p>
"
`;

exports[`caniusePlugin > should not work 4`] = `
"<p>@[caniuse](xxx</p>
"
`;

exports[`caniusePlugin > should not work 5`] = `
"<p>@[caniuse](</p>
"
`;

exports[`caniusePlugin > should not work 6`] = `
"<p>@[caniuse</p>
"
`;

exports[`caniusePlugin > should not work 7`] = `
"<p>@<a href="">caniuse)</a></p>
"
`;

exports[`caniusePlugin > should not work 8`] = `
"<p>@<a href="">caniuseee)</a></p>
"
`;

exports[`caniusePlugin > should work 1`] = `"<CanIUseViewer feature="feature" meta="test-id" past="2" future="1" />"`;

exports[`caniusePlugin > should work 2`] = `
"<ClientOnly><p><picture>
      <source type="image/webp" srcset="https://caniuse.bitsofco.de/image/feature.webp">
      <source type="image/png" srcset="https://caniuse.bitsofco.de/image/feature.png">
      <img src="https://caniuse.bitsofco.de/image/feature.jpg" alt="Data on support for the feature feature across the major browsers from caniuse.com" width="100%">
    </picture></p></ClientOnly>"
`;

exports[`caniusePlugin > should work 3`] = `"<CanIUseViewer feature="feature" meta="test-id" past="2" future="1" />"`;

exports[`caniusePlugin > should work 4`] = `"<CanIUseViewer feature="feature" meta="test-id" past="2" future="0" />"`;

exports[`caniusePlugin > should work 5`] = `"<CanIUseViewer feature="feature" meta="test-id" past="2" future="0" /><CanIUseViewer feature="feature" meta="test-id" past="2" future="0" />"`;

exports[`caniusePlugin > should work 6`] = `"<CanIUseViewer feature="feature" meta="test-id" past="2" future="0" />"`;

exports[`caniusePlugin > should work with options 1`] = `
"<ClientOnly><p><picture>
      <source type="image/webp" srcset="https://caniuse.bitsofco.de/image/feature.webp">
      <source type="image/png" srcset="https://caniuse.bitsofco.de/image/feature.png">
      <img src="https://caniuse.bitsofco.de/image/feature.jpg" alt="Data on support for the feature feature across the major browsers from caniuse.com" width="100%">
    </picture></p></ClientOnly>"
`;

exports[`caniusePlugin > should work with options 2`] = `"<CanIUseViewer feature="feature" meta="test-id" past="2" future="1" />"`;

exports[`caniusePlugin > should work with options 3`] = `"<CanIUseViewer feature="feature" meta="test-id" past="2" future="0" />"`;

exports[`legacyCaniuse > should work 1`] = `"<CanIUseViewer feature="feature" meta="test-id" past="2" future="1" />"`;

exports[`legacyCaniuse > should work 2`] = `"<CanIUseViewer feature="feature{-2,4}" meta="test-id" past="2" future="0" />"`;

exports[`legacyCaniuse > should work 3`] = `""`;

exports[`legacyCaniuse > should work with unknown mode 1`] = `
"<ClientOnly><p><picture>
      <source type="image/webp" srcset="https://caniuse.bitsofco.de/image/feature.webp">
      <source type="image/png" srcset="https://caniuse.bitsofco.de/image/feature.png">
      <img src="https://caniuse.bitsofco.de/image/feature.jpg" alt="Data on support for the feature feature across the major browsers from caniuse.com" width="100%">
    </picture></p></ClientOnly>"
`;

exports[`legacyCaniuse > should work with unknown mode 2`] = `
"<ClientOnly><p><picture>
      <source type="image/webp" srcset="https://caniuse.bitsofco.de/image/feature{-2,4}.webp">
      <source type="image/png" srcset="https://caniuse.bitsofco.de/image/feature{-2,4}.png">
      <img src="https://caniuse.bitsofco.de/image/feature{-2,4}.jpg" alt="Data on support for the feature{-2,4} feature across the major browsers from caniuse.com" width="100%">
    </picture></p></ClientOnly>"
`;
